dask/config.py

```diff
@@ -0,0 +1,304 @@
+        source,
+        destination=os.path.join(os.path.expanduser('~'), '.config', 'dask'),
+        comment=True):
+    """
+    Copy file to default location if it does not already exist
+
+    This tries to move a default configuration file to a default location if
+    if does not already exist.  It also comments out that file by default.
+
+    This is to be used by downstream modules (like dask.distributed) that may
+    have default configuration files that they wish to include in the default
+    configuration path.
+
+    Parameters
+    ----------
+    source: string, filename
+        source configuration file, typically within a source directory
+    destination: string, filename
+        destination filename, typically ~/.config/dask
+    comment: bool, True by default
+        Whether or not to comment out the config file when copying
+    """
+    if not os.path.splitext(destination)[1].strip('.'):
+        _, filename = os.path.split(source)
+        destination = os.path.join(destination, filename)
+
+    if not os.path.exists(os.path.dirname(destination)):
+        os.makedirs(os.path.dirname(destination))
+
+    if not os.path.exists(destination):
+        # Atomically create destination.  Parallel testing discovered
+        # a race condition where a process can be busy creating the
+        # destination while another process reads an empty config file.
+        tmp = '%s.tmp.%d' % (destination, os.getpid())
+        with open(source) as f:
+            lines = list(f)
+
+        if comment:
+            lines = ['# ' + line
+                     if line.strip() and not line.startswith('#')
+                     else line
+                     for line in lines]
+
+        with open(tmp, 'w') as f:
+            f.write(''.join(lines))
+
+        try:
+            os.rename(tmp, destination)
+        except OSError:
+            os.remove(tmp)
+
+
+class set(object):
+    """ Temporarily set configuration values within a context manager
+
+    Examples
+    --------
+    >>> import dask
+    >>> with dask.config.set({'foo': 123}):
+    ...     pass
+
+    See Also
+    --------
+    dask.config.get
+    """
+    def __init__(self, arg=None, config=None, **kwargs):
+        if config is None:
+            config = global_config
+        if arg and not kwargs:
+            kwargs = arg
+
+        self.config = config
+        self.old = copy.deepcopy(config)
+
+        def assign(keys, value, d):
+            key = keys[0]
+            if len(keys) == 1:
+                d[keys[0]] = value
+            else:
+                if key not in d:
+                    d[key] = {}
+                assign(keys[1:], value, d[key])
+
+        for key, value in kwargs.items():
+            assign(key.split('.'), value, config)
+
+    def __enter__(self):
+        return config
+
+    def __exit__(self, type, value, traceback):
+        self.config.clear()
+        self.config.update(self.old)
+
+
+configs = []
+
+try:
+    import yaml
```

Comment:
This logic feels a bit off to me, as functions rely on yaml being defined globally, but the import is hidden in the middle of the file. I'd prefer to see something like:

```
try:
    import yaml
except ImportError:
    yaml = None

...

if yaml is None:
    configs.extend(collect_yaml())
```

Reply:
Fixed